pkg/gardenlet/controller/shoot/shoot/cleaner.go

```diff
@@ -103,11 +103,15 @@ func (c *cleaner) WaitUntilExtensionObjectsDeleted(ctx context.Context) error {
 }
 
 // DeleteMachineResources deletes all MachineControllerManager resources in the shoot namespace.
+// The resources are force deleted sequentially in the order MachineDeployment, MachineSet, Machine, MachineClass to prevent cases
+// where Machines are force deleted first and then machine-controller-manager Pod creates a new Machine as replacement.
 func (c *cleaner) DeleteMachineResources(ctx context.Context) error {
-	return utilclient.ApplyToObjectKinds(ctx, func(kind string, objectList client.ObjectList) flow.TaskFn {
-		c.log.Info("Deleting all machine resources in namespace", "namespace", c.seedNamespace, "kind", kind)
-		return utilclient.ForceDeleteObjects(c.seedClient, c.seedNamespace, objectList)
-	}, machineKindToObjectList)
+	return flow.Sequential(
+		utilclient.ForceDeleteObjects(c.seedClient, c.seedNamespace, &machinev1alpha1.MachineDeploymentList{}),
```

Comment:
Should we include the wait here? For resources to be gone. 
Can it happen if the deployment has deletion time stamp still machines can be created for it? (Mostly not). If not then it's fine.

Reply:
Also, please adapt https://github.com/gardener/gardener/blob/c29e9754aaea33814d74447601702e5f1a9740b9/pkg/gardenlet/controller/shoot/shoot/cleaner.go#L56-L61 list in same as new order.

Reply:
> Should we include the wait here? For resources to be gone.
> Can it happen if the deployment has deletion time stamp still machines can be created for it? (Mostly not). If not then it's fine.

In the flows, we seem to do actions in tasks and then wait for conditions in other tasks. You rather prefer the force deletion and wait to be sequential - for example first force delete MachineDeployments, then wait for them to be gone, only then proceed with the MachineSets.
If you want, I can change the implementation but I assume that If MachineDeployment (or MachineSet) has a deletionTimestamp, then it will only try to delete (and not create) the corresponding MachineSet (or Machine). MCM folks (@himanshu-kun , @rishabh-11), can you  confirm?

Reply:
yes it'll try to delete the child machinesets https://github.com/gardener/machine-controller-manager/blob/1b57e16c5c3df14588a88399406d57a2a7ab3008/pkg/controller/deployment.go#L495-L506